packages/runtime/container-runtime/src/test/throttler.spec.ts

```diff
@@ -134,4 +126,202 @@ describe("Throttler", () => {
+        // 11 | 2^(10 - 1) x 20 =  512 x 20 = | 10,240 ms | 20,460 ms
+        // 12 | 2^(11 - 1) x 20 = 1024 x 20 = | 20,480 ms | 40,940 ms
+        // 13 | 2^(12 - 1) x 20 = 2048 x 20 = | 30,000 ms | 70,940 ms (MAX)
+        // 14 | 2^(13 - 1) x 20 = 5096 x 20 = | 30,000 ms |100,940 ms (MAX)
+        const delayFn = formExponentialFnWithAttemptOffset(-1, {
+            coefficient: 20,
+            initialDelay: 0,
+        });
+
+        beforeEach(() => throttler = new Throttler(delayWindowMs, maxDelayMs, delayFn));
+
+        it("Should initially have zero delay", () => {
+            assert.strictEqual(throttler.getDelay(), 0);
+        });
+
+        it("Should increase as expected with instant failures", () => {
+            const expectedDelays = [
+                0, 20, 40, 80, 160, 320, 640,
+                1280, 2560, 5120, 10240, 20480,
+                30000, 30000, 30000, 30000, 30000,
+            ];
+            for (const expectedDelay of expectedDelays) {
+                assert.strictEqual(getDelayAndTick(), expectedDelay);
+            }
+        });
+
+        it("Should remain zero delay with long pauses between getDelay calls", () => {
+            for (let i = 0; i < 5; i++) {
+                assert.strictEqual(getDelayAndTick(), 0, `iteration ${i}`);
+                clock.tick(delayWindowMs);
+            }
+            assert.strictEqual(getDelayAndTick(), 0);
+
+            // This time barely keep it in the window, giving a delay.
+            clock.tick(delayWindowMs - 1);
+            assert.strictEqual(getDelayAndTick(), 20);
+        });
+
+        it("Should not increase with long pauses between getDelay calls", () => {
+            const oneThirdTicks = Math.floor(delayWindowMs / 3);
+            const remainingTicks = delayWindowMs - (2 * oneThirdTicks);
+
+            // Accumulate some attempts first.
+            assert.strictEqual(getDelayAndTick(), 0);
+            clock.tick(oneThirdTicks);
+            assert.strictEqual(getDelayAndTick(), 20);
+            clock.tick(oneThirdTicks);
+            assert.strictEqual(getDelayAndTick(), 40);
+            clock.tick(remainingTicks);
+
+            // Loop through attempts periodically dropping off.
+            for (let i = 0; i < 100; i++) {
+                assert.strictEqual(getDelayAndTick(), 40, `iteration ${i}`);
+                clock.tick(i % 3 === 2 ? remainingTicks : oneThirdTicks);
+            }
+            assert.strictEqual(getDelayAndTick(), 40);
+
+            // This time fail instantly, giving a later delay.
+            assert.strictEqual(getDelayAndTick(), 80);
+        });
+
+        it("Should stop increasing number of attempts after max", () => {
+            for (let i = 0; i < 12; i++) {
+                getDelayAndTick();
+                assert.strictEqual(throttler.numAttempts, i + 1, `loop 1; iteration ${i}`);
+            }
+            for (let i = 0; i < 100; i++) {
```

Comment:
Most likely I'm advocating for some premature optimization wrt to how long it takes to run the tests :)

But do we really need to go to 100 in this loop? This is traversing an array (small one though) 100 times, right?

Reply:
It probably doesn't need to go to 100.. I was thinking about potential edge cases: off by 1, where it slides one attempt out of the window after many attempts. I didn't actually calculate it out though, which means that 100 is very arbitrary, and maybe not even sufficient.